modules/function/src/main/java/com/opengamma/strata/function/calculation/deposit/TermDepositCalculationFunction.java

```diff
@@ -0,0 +1,129 @@
+import com.opengamma.strata.market.key.DiscountCurveKey;
+import com.opengamma.strata.product.deposit.ExpandedTermDeposit;
+import com.opengamma.strata.product.deposit.TermDeposit;
+import com.opengamma.strata.product.deposit.TermDepositTrade;
+
+/**
+ * Perform calculations on a single {@code TermDepositTrade} for each of a set of scenarios.
+ * <p>
+ * This uses the standard discounting calculation method.
+ * The supported built-in measures are:
+ * <ul>
+ *   <li>{@linkplain Measure#PAR_RATE Par rate}
+ *   <li>{@linkplain Measure#PAR_SPREAD Par spread}
+ *   <li>{@linkplain Measure#PRESENT_VALUE Present value}
+ *   <li>{@linkplain Measure#PV01 PV01}
+ *   <li>{@linkplain Measure#BUCKETED_PV01 Bucketed PV01}
+ * </ul>
+ */
+public class TermDepositCalculationFunction
+    implements CalculationMultiFunction<TermDepositTrade> {
+
+  /**
+   * The calculations by measure.
+   */
+  private static final ImmutableMap<Measure, SingleMeasureCalculation> CALCULATORS =
+      ImmutableMap.<Measure, SingleMeasureCalculation>builder()
+          .put(Measure.PAR_RATE, TermDepositMeasureCalculations::parRate)
+          .put(Measure.PAR_SPREAD, TermDepositMeasureCalculations::parSpread)
+          .put(Measure.PRESENT_VALUE, TermDepositMeasureCalculations::presentValue)
+          .put(Measure.PV01, TermDepositMeasureCalculations::pv01)
+          .put(Measure.BUCKETED_PV01, TermDepositMeasureCalculations::bucketedPv01)
+          .build();
+
+  /**
+   * Creates an instance.
+   */
+  public TermDepositCalculationFunction() {
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  public Set<Measure> supportedMeasures() {
+    return CALCULATORS.keySet();
+  }
+
+  @Override
+  public Optional<Currency> defaultReportingCurrency(TermDepositTrade target) {
+    return Optional.of(target.getProduct().getCurrency());
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  public FunctionRequirements requirements(TermDepositTrade trade, Set<Measure> measures) {
+    TermDeposit product = trade.getProduct();
+
+    Set<DiscountCurveKey> discountCurveKeys =
+        ImmutableSet.of(DiscountCurveKey.of(product.getCurrency()));
+
+    return FunctionRequirements.builder()
+        .singleValueRequirements(discountCurveKeys)
+        .timeSeriesRequirements()
+        .outputCurrencies(product.getCurrency())
+        .build();
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  public Map<Measure, Result<ScenarioResult<?>>> calculate(
+      TermDepositTrade trade,
+      Set<Measure> measures,
+      CalculationMarketData scenarioMarketData) {
+
+    // expand the trade once for all measures and all scenarios
+    ExpandedTermDeposit product = trade.getProduct().expand();
+
+    // loop around measures, calculating all scenarios for one measure
+    Map<Measure, Result<ScenarioResult<?>>> results = new HashMap<>();
```

Comment:
Deliberately chose to make the function unabstracted and with a minimum of streams. A key problem in the past has been getting people to debug in, step through and see where things are going wrong. By making the calculate method of these function classes as simple as possible, using just classic Java, it will help them debug. Abstractions in particular seem to be part of the problem in functions.

FWIW, not all of the interfaces take three arguments. Some trade types need 2 and CDS needs more.

I did consider making the `*MeasureCalculations` classes also avoid streams, but as it stands the collector is too useful there. I have fixed the code to avoid boxing in parRate/parSpread.